tsdb/chunks/head_chunks.go

```diff
@@ -0,0 +1,753 @@
+	}
+
+	return nil
+}
+
+func sequenceFilesMap(dir string) (map[int]string, error) {
+	files, err := ioutil.ReadDir(dir)
+	if err != nil {
+		return nil, err
+	}
+	res := map[int]string{}
+	for _, fi := range files {
+		seq, err := strconv.ParseUint(fi.Name(), 10, 64)
+		if err != nil {
+			continue
+		}
+		res[int(seq)] = filepath.Join(dir, fi.Name())
+	}
+	return res, nil
+}
+
+// WriteChunk writes the chunk to the disk.
+// The returned chunk ref is the reference from where the chunk encoding starts for the chunk.
+func (cdm *ChunkDiskMapper) WriteChunk(seriesRef uint64, mint, maxt int64, chk chunkenc.Chunk) (chunkRef uint64, err error) {
+	cdm.wbufLock.Lock()
+	defer cdm.wbufLock.Unlock()
+
+	if cdm.closed {
+		return 0, ErrChunkDiskMapperClosed
+	}
+
+	if cdm.shouldCutNewFile(len(chk.Bytes()), maxt) {
+		if err := cdm.cut(mint); err != nil {
+			return 0, err
+		}
+		cdm.curFileMaxt = 0
+	}
+
+	// seriesRef(8) + mint(8) + maxt(8) + encoding(1) + max uvarint size (5) + the bytes is
+	// the total size to be written.
+	// if len(chk.Bytes()) >= writeBufferSize-30, it means that chunk >= the buffer size;
+	// so no need to flush here, as we have to flush at the end (to not keep partial chunks in buffer).
+	if len(chk.Bytes()) < writeBufferSize-30 && cdm.wbuf.Available() < 8+8+8+1+5+len(chk.Bytes()) {
+		if err := cdm.flushBuffer(); err != nil {
+			return 0, err
+		}
+	}
+
+	cdm.crc32.Reset()
+	binary.BigEndian.PutUint64(cdm.writeBuf[:], seriesRef)
+	if err := cdm.writeWithCRC32(cdm.writeBuf[:SeriesRefSize]); err != nil {
+		return 0, err
+	}
+
+	binary.BigEndian.PutUint64(cdm.writeBuf[:], uint64(mint))
+	if err := cdm.writeWithCRC32(cdm.writeBuf[:MintMaxtSize]); err != nil {
+		return 0, err
+	}
+
+	binary.BigEndian.PutUint64(cdm.writeBuf[:], uint64(maxt))
+	if err := cdm.writeWithCRC32(cdm.writeBuf[:MintMaxtSize]); err != nil {
+		return 0, err
+	}
+
+	// The reference is set to the head chunk file index and the offset where
+	// the data starts for this chunk.
+	//
+	// The upper 4 bytes are for the head chunk file index and
+	// the lower 4 bytes are for the head chunk file offset where to start reading this chunk.
+	chunkRef = cdm.chunkRef(uint64(cdm.curFileSequence), uint64(cdm.n))
+
+	cdm.writeBuf[0] = byte(chk.Encoding())
+	if err := cdm.writeWithCRC32(cdm.writeBuf[:1]); err != nil {
+		return 0, err
+	}
+
+	n := binary.PutUvarint(cdm.writeBuf[:], uint64(len(chk.Bytes())))
+	if err := cdm.writeWithCRC32(cdm.writeBuf[:n]); err != nil {
+		return 0, err
+	}
+
+	if err := cdm.writeWithCRC32(chk.Bytes()); err != nil {
+		return 0, err
+	}
+
+	if err := cdm.write(cdm.crc32.Sum(cdm.writeBuf[:0])); err != nil {
+		return 0, err
+	}
+
+	if maxt > cdm.curFileMaxt {
+		cdm.curFileMaxt = maxt
+	}
+
+	if mint < cdm.curFileMint {
+		cdm.curFileMint = mint
+		// As this wont happen a whole lot of time, we don't wait
+		// for a new head chunk file to be cut to write it off.
+		binary.BigEndian.PutUint64(cdm.writeBuf[:], uint64(mint))
```

Comment:
Why it matters? Can we just stick to `cut` and have all in single place for simplicity?

Reply:
Maxt depends on the chunks. We cannot have the maxt of all the chunks to be written to the file in future in `cut` :). So we update the maxt whenever we write chunks.

Reply:
We need maxt for the truncation of files. (Although you could argue that you can get away with mint)

Reply:
Turns out I gave a wrong explanation here. So we write the maxt only at the end, and if Prometheus happens to crash, maxt remains 0 in the file. So we are maintaining mint in on the disk, which is checked at every WriteChunk. Using this mint, we can optimistically set the maxt for the crashed file when we are getting Prometheus back up (to help with the truncation).

Reply:
Another thing we could do is, when we get a new chunk to write after Prometheus comes back, we can use it's mint as maxt for all the old files which have 0 as maxt (which means crash). This could mean a delayed deletion of files (which might hurt the actual data retention?).